### PR TITLE
feat: repository and templates client api

### PR DESCRIPTION
**Pull request by lkingland**
# Changes

- :gift: Templates management api
- :gift: Repositories management api expansion

Prerequisites for #453 and #451

/kind enhancement

### CODE REVIEW

File: repository.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+// runtimes
+func NewRepositoryFromPath(path string) (Repository, error) {
+	// TODO: read and use manifest if it exists
+
+	r := Repository{
+		Name:      filepath.Base(path),
+		Templates: []Template{},
+		Runtimes:  []string{}}
+
+	// Each subdirectory is a Runtime
```
Comment (lance) [inline repository.go:37]: I suppose a `manifest.yaml` might specify either includes/excludes or have some other mechanism to make sure, we don't end up displaying `test` or something that's not actually a runtime dir?

File: repositories_test.go [hunk 2]
```diff
@@ -13,51 +12,110 @@ import (
...
 
 	rr, err := client.Repositories.List()
 	if err != nil {
 		t.Fatal(err)
 	}
-	if len(rr) != 0 {
-		t.Fatalf("Expected an empty repositories list, got %v", len(rr))
+	// Assert contains only the default repo
+	if len(rr) != 1 && rr[0] != fn.DefaultRepository {
+		t.Fatalf("Expected repositor list '[%v]', got %v", fn.DefaultRepository, rr)
```
Comment (lance) [inline repositories_test.go:29]: Nit
```suggestion
		t.Fatalf("Expected repository list '[%v]', got %v", fn.DefaultRepository, rr)
```

File: repositories_test.go [hunk 3]
```diff
@@ -13,51 +12,110 @@ import (
...
+
+	// valid should not error
+	repo, err = client.Repositories.Get("customProvider")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// valid should have expected name
+	if repo.Name != "customProvider" {
+		t.Fatalf("Repository returnd expected 'customProvider', got: %v", repo.Name)
```
Comment (lance) [inline repositories_test.go:51]: Nit
```suggestion
		t.Fatalf("expected 'customProvider' as repository name, got: %v", repo.Name)
```

**Review by lance — COMMENTED** [review]
Just a few small comments. Otherwise this looks great! Apologies for the latency in reviewing it.

/approve

File: repository.go (hunk 1 above)
Comment (lkingland) [reply repository.go:37]: Exactly what I was thinking as well.  Directory structure a sensible default, `manifset.yaml` an optional way to override things (how many things and exactly _how_ ?  That's still a bit mind-murky)

File: repositories.go [hunk 4]
```diff
@@ -10,10 +10,69 @@ import (
...
+	}
+	return names, nil
+}
+
+// All repositories under management (at configured Path)
+func (r *Repositories) All() (repos []Repository, err error) {
+	repos = []Repository{}
+
+	// Single repo override
+	// TODO: Create single remote repository override for WithRepository option.
```
[the remark this answers is not in the document]
Comment (lkingland) [reply repositories.go:37]: Next on the list!

File: client.go [hunk 5]
```diff
@@ -604,6 +623,45 @@ func (c *Client) Emit(ctx context.Context, endpoint string) error {
...
+
+func (s *sortedSet) Add(value string) {
+	s.Lock()
+	s.members[value] = true
+	s.Unlock()
+}
+
+func (s *sortedSet) Remove(value string) {
+	s.Lock()
+	s.members[value] = false
```
[the remark this answers is not in the document]
Comment (lkingland) [reply client.go:650]: Ugh; thank you for noticing.  That method was being used only for some debugging and is now removed entirely.

**Comment by lkingland** [comment]
Thank you very much @lance for the review!  This one is a bit of a read.
If it passes muster now, we'll just need a consent for the googlebot and an /lgtm label to merge

**Comment by lance** [comment]
> we'll just need a consent for the googlebot

￼I think this might happen when accepting suggestions in the GitHub UI. A rebase squashing the commit should fix it.

**Comment by lance** [comment]
@googlebot I consent.

**Pull request merged**